R/parse-body.R

```diff
@@ -16,230 +12,461 @@ postBodyFilter <- function(req){
+        # (Ex: `.rds` files -> `application/rds` which has a proper RDS parser)
+        isTRUE(stri_detect_fixed(x$content_type, "application/octet-stream"))
+      ) {
+        if (!is.null(x$filename)) {
+          # Guess content-type from file extension
+          x$content_type <- getContentType(tools::file_ext(x$filename))
+        }
       }
-      parseRaw(x)
+      x$parsers <- parsers
+      parse_raw(x)
     })
   }
 }
 
-
-
-
-#' OCTET
-#' @rdname parsers
+#' @describeIn parsers Enable all parsers. Not recommended due to security concerns.
 #' @export
-parser_octet <- function() {
-  function(value, filename = NULL, ...) {
-    attr(value, "filename") <- filename
-    return(value)
+parser_all <- function() {
+  function(value, content_type, filename, ...) {
+    # re-perform parse_raw(), but provide all parsers
+    parse_raw(
+      # mimic the shape of the output of `webutils::parse_multipart` + parsers
+      list(
+        content_type = content_type,
+        value = value,
+        filename = filename,
+        parsers = .globals$parsers
+      )
+    )
   }
 }
 
-
-
-
-#' YAML
-#' @rdname parsers
+#' @describeIn parsers No parser. Will not process the postBody.
 #' @export
-parser_yaml <- function() {
-  if (!requireNamespace("yaml", quietly = TRUE)) {
-    stop("yaml must be installed for the yaml parser to work")
-  }
-  function(value, content_type = NULL, ...) {
-    charset <- getCharacterSet(content_type)
-    value <- rawToChar(value)
-    Encoding(value) <- charset
-    yaml::yaml.load(value)
+parser_none <- function() {
+  function(value, ...) {
+    value
   }
 }
 
-
-
-
-addParsers_onLoad <- function() {
-  addParser("json", parser_json, "application/json")
-  addParser("query", parser_query, "application/x-www-form-urlencoded")
-  addParser("text", parser_text, "text/")
-  addParser("rds", parser_rds, "application/rds")
-  addParser("multi", parser_multi, "multipart/form-data")
-  addParser("octet", parser_octet, "application/octet")
-  addParser("yaml", parser_yaml, "application/x-yaml")
+register_parsers_onLoad <- function() {
+  # parser alias names for plumbing
+  register_parser("csv", parser_csv, fixed = c("application/csv", "application/x-csv", "text/csv", "text/x-csv"))
+  register_parser("json", parser_json, fixed = c("application/json", "text/json"))
+  register_parser("multi", parser_multi, fixed = "multipart/form-data")
+  register_parser("octet", parser_octet, fixed = "application/octet-stream")
+  register_parser("query", parser_query, fixed = "application/x-www-form-urlencoded")
+  register_parser("rds", parser_rds, fixed = "application/rds")
+  register_parser("text", parser_text, fixed = "text/plain", regex = "^text/")
+  register_parser("tsv", parser_tsv, fixed = c("application/tab-separated-values", "text/tab-separated-values"))
+  register_parser("yaml", parser_yaml, fixed = c("application/yaml", "application/x-yaml", "text/yaml", "text/x-yaml"))
```

Comment:
Maybe we should not register yaml parser since it requires yaml library. Using the all parser could lead to a yaml library error.

Reply:
Moved the library check to inside the parsing function.